DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/register_map_and_streaming_interfaces/src/register_map_functor_model.cpp

```diff
@@ -0,0 +1,118 @@
+#include <sycl/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/prototype/interfaces.hpp>
+
+#include "exception_handler.hpp"
+
+using namespace sycl;
+
+using ValueT = int;
+
+// offloaded computation
+ValueT SomethingComplicated(ValueT val) { return (ValueT)(val * (val + 1)); }
+
+/////////////////////////////////////////
+
+struct FunctorRegisterMapIP {
+  // Use the 'register_map' annotation on a kernel argument to specify it to be
+  // a register map kernel argument.
+  register_map ValueT *input;
+  // Without the annotations, kernel arguments will be inferred to be register
+  // map kernel arguments if the kernel invocation interface is register mapped,
+  // and vise-versa.
+  ValueT *output;
+  // A kernel with a register map invocation interface can also independently
+  // have streaming kernel arguments, when annotated by 'conduit'.
+  conduit size_t n;
+  register_map_interface void operator()() const {
+    for (int i = 0; i < n; i++) {
+      output[i] = SomethingComplicated(input[i]);
+    }
+  }
+};
+
+int main(int argc, char *argv[]) {
+#if FPGA_SIMULATOR
+  auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+  auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+  auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+  bool passed = true;
+
+  size_t count = 16;
+  if (argc > 1)
+    count = atoi(argv[1]);
+
+  if (count <= 0) {
+    std::cerr << "ERROR: 'count' must be positive" << std::endl;
+    return 1;
+  }
+
+  try {
+    // create the device queue
+    sycl::queue q(selector, fpga_tools::exception_handler);
+
+    // make sure the device supports USM host allocations
+    sycl::device d = q.get_device();
```

Comment:
Please add this to all the source codes: 
    // Print out the device information.
    std::cout << "Running on device: "
              << q.get_device().get_info<info::device::name>() << "\n";